src/charm.py

```diff
@@ -1705,7 +1705,46 @@ def _setup_ldap_sync(self, postgres_snap: snap.Snap | None = None) -> None:
         logger.debug("Starting LDAP sync service")
         postgres_snap.restart(services=["ldap-sync"])
 
-    def _start_primary(self, event: StartEvent) -> None:  # noqa: C901
+    def _setup_users(self) -> None:
+        self.postgresql.create_predefined_instance_roles()
+
+        # Create the default postgres database user that is needed for some
+        # applications (not charms) like Landscape Server.
+
+        # This event can be run on a replica if the machines are restarted.
+        # For that case, check whether the postgres user already exits.
+        users = self.postgresql.list_users()
+        # Create the backup user.
+        if BACKUP_USER not in users:
+            self.postgresql.create_user(
+                BACKUP_USER, new_password(), extra_user_roles=[ROLE_BACKUP]
+            )
+            self.postgresql.grant_database_privileges_to_user(BACKUP_USER, "postgres", ["connect"])
+        if MONITORING_USER not in users:
+            # Create the monitoring user.
+            self.postgresql.create_user(
+                MONITORING_USER,
+                self.get_secret(APP_SCOPE, MONITORING_PASSWORD_KEY),
+                extra_user_roles=[ROLE_STATS],
+            )
+
+        self.postgresql.set_up_database(
+            temp_location="/var/snap/charmed-postgresql/common/data/temp"
+        )
+
+        access_groups = self.postgresql.list_access_groups()
+        if access_groups != set(ACCESS_GROUPS):
+            self.postgresql.create_access_groups()
+            self.postgresql.grant_internal_access_group_memberships()
+
+        access_groups = self.postgresql.list_access_groups()
+        if access_groups != set(ACCESS_GROUPS):
+            self.postgresql.create_access_groups()
+            self.postgresql.grant_internal_access_group_memberships()
```

Comment:
This block can be removed (it was already duplicated in the charm code before this PR).